lib/src/edit_builder.dart

```diff
@@ -0,0 +1,869 @@
+
+  @override
+  String toString() => _yaml;
+
+  YamlEditBuilder(this._yaml, {this.indentationStep = 2}) {
+    var contents = loadYamlNode(_yaml);
+    _contents = _modifiedYamlNodeFrom(contents, this, []);
+  }
+
+  /// Traverses down the provided [path] to the _ModifiableYamlNode at [path].
+  _ModifiableYamlNode _traverse(Iterable<Object> path) {
+    var current = _contents;
+    for (var key in path) {
+      current = current[key];
+    }
+    return current;
+  }
+
+  /// Traverses down the provided [path] to the _ModifiableYamlList at [path].
+  _ModifiableYamlList _traverseToList(Iterable<Object> path) {
+    var possibleList = _traverse(path);
+
+    if (possibleList is _ModifiableYamlList) {
+      return possibleList;
+    } else {
+      throw TypeError();
+    }
+  }
+
+  /// Returns the value of the node at path as a YamlNode.
+  YamlNode parseValueAt(Iterable<Object> path) {
+    var value = _traverse(path).value;
+
+    if (value is _ModifiableYamlList) {
+      return YamlList.wrap(value);
+    } else if (value is _ModifiableYamlMap) {
+      return YamlMap.wrap(value);
+    } else {
+      return YamlScalar.wrap(value);
+    }
+  }
+
+  /// Sets [value] in the [path]. If the [path] is not accessible (e.g. it currently
+  /// does not exist in the document), an error will be thrown.
+  void setIn(Iterable<Object> path, Object value) {
+    var yamlCollection = _traverse(path.take(path.length - 1));
+
+    if (yamlCollection is _ModifiableYamlList) {
+      var lastNode = path.last;
+      yamlCollection[lastNode] = value;
+    } else if (yamlCollection is _ModifiableYamlMap) {
+      var lastNode = path.last;
+      yamlCollection[lastNode] = value;
+    } else {
+      throw TypeError();
+    }
+  }
+
+  /// Appends [value] into the list at [listPath], only if the element at the given path
+  /// is a List.
+  void addInList(Iterable<Object> listPath, Object value) {
+    var yamlList = _traverseToList(listPath);
+    yamlList.add(value);
+  }
+
+  /// Prepends [value] into the list at [listPath], only if the element at the given path
+  /// is a List.
+  void prependInList(Iterable<Object> listPath, Object value) {
+    var yamlList = _traverseToList(listPath);
+    yamlList.prepend(value);
+  }
+
+  /// Inserts [value] into the list at [listPath], only if the element at the given path
+  /// is a list. [index] must be non-negative and no greater than the list's length.
+  void insertInList(Iterable<Object> listPath, int index, Object value) {
+    var yamlList = _traverseToList(listPath);
+    yamlList.insert(index, value);
+  }
+
+  /// Removes the value in the path.
+  void removeIn(Iterable<Object> path) {
+    var current = _traverse(path.take(path.length - 1));
+    var lastNode = path.last;
+
+    if (current is _ModifiableYamlList) {
+      current.removeAt(lastNode);
+    } else if (current is _ModifiableYamlMap) {
+      current.remove(lastNode);
+    } else {
+      throw TypeError();
```

Comment:
Isn't the problem here to do with the `path` that was given as input?

Maybe we should make a `PathError` that extends `RangeError` from the core library? Or we could just make a `RangeError` instance and use.

Note. I think throwing an error instead of an exception is reasonable in these mutation methods. Because the programmer should have checked that the mutation was possible, rather than try to do the mutation and handle the exception.

Reply:
Will fix after trying to remove `_ModifiableYamlNode`!